File: unit-testing-workshop/test/com/ihsinformatics/unittesting/DateTimeUtilTest.java
/**
 * 
 */
package com.ihsinformatics.unittesting;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNotSame;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.security.InvalidParameterException;
import java.util.Arrays;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.junit.Test;

/**
 * @author owais.hussain@ihsinformatics.com
 *
 */
public class DateTimeUtilTest {

	String sqlFormat = "yyyy-MM-dd HH:mm:ss";
	String isoFormat = "yyyy-MM-dd'T'HH:mm:ss.'Z'";
	String clientFormat = "dd/MM/yyyy HH:mm";

	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#fromString(java.lang.String, java.lang.String)}.
	 */
	@Test
	public void shouldConvertStringDateInClientFormat() {
		Date date = DateTimeUtil.fromString("01/01/2018 00:00", clientFormat);
		assertNotNull(date);
	}

	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#fromString(java.lang.String, java.lang.String)}.
	 */
	@Test
	public void shouldConvertStringDateTimeInClientFormat() {
		Date date = DateTimeUtil.fromString("01/01/2018 16:55", clientFormat);
		assertNotNull(date);
	}

	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#fromString(java.lang.String, java.lang.String)}.
	 */
	@Test
	public void shouldConvertUnusualStringDatesInClientFormat() {
		List<String> list = Arrays.asList("-1/01/2018 00:00", "32/01/2018 00:00", "01/13/2018 00:00",
				"01/01/-800 00:00", "01/01/2018 24:00", "01/01/2018 00:60");
		list.forEach(dateString -> assertNotNull("Should not convert " + dateString,
				DateTimeUtil.fromString(dateString, clientFormat)));
	}

	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#fromString(java.lang.String, java.lang.String)}.
	 */
	@Test
	public void testAmbiguousClientFormatDatesFromString() {
		Date date1 = DateTimeUtil.fromString("12/05/2018 00:00", clientFormat);
		Date date2 = DateTimeUtil.fromString("05/12/2018 00:00", clientFormat);
		assertTrue(date1.before(date2));
	}

	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#toString(java.util.Date, java.lang.String)}.
	 */
	@SuppressWarnings("deprecation")
	@Test
	public void shouldConvertDatesToClientFormatString() {
		Map<Date, String> dates = new HashMap<>();
		dates.put(new Date(118, 4, 26, 0, 0), "26/05/2018 00:00");
		dates.put(new Date(118, 4, 26, 12, 30), "26/05/2018 12:30");
		dates.put(new Date(100, 0, 1, 0, 0), "01/01/2000 00:00");
		dates.forEach((date, dateString) -> assertEquals(DateTimeUtil.toString(date, clientFormat), dateString));
	}

	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#fromSqlDateString(java.lang.String)}.
	 */
	@Test
	public void testFromSqlDateString() {
		fail("Not yet implemented");
	}

	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#fromSqlDateTimeString(java.lang.String)}.
	 */
	@Test
	public void testFromSqlDateTimeString() {
		fail("Not yet implemented");
	}

	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#toSqlDateString(java.util.Date)}.
	 */
	@Test
	public void testToSqlDateString() {
		fail("Not yet implemented");
	}

	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#toSqlDateTimeString(java.util.Date)}.
	 */
	@Test
	public void testToSqlDateTimeString() {
		fail("Not yet implemented");
	}

	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#detectDateFormat(java.lang.String)}.
	 */
	@Test(expected = InvalidParameterException.class)
	public void shouldThrowInvalidParameterExceptionOnUnknownFormat() {
		String dateString = "14081947 010000";
		DateTimeUtil.detectDateFormat(dateString);
	}
	
	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#detectDateFormat(java.lang.String)}.
	 */
	@Test
	public void shouldDetectSQLDateFormat() {
		String dateString = "1947-08-14 01:00:00";
		String detectedFormat = DateTimeUtil.detectDateFormat(dateString);
		assertSame(detectedFormat, sqlFormat);
	}


	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#detectDateFormat(java.lang.String)}.
	 */
	@Test
	public void shouldDetectClientDateFormat() {
		String dateString = "14/08/1947 01:00";
		String detectedFormat = DateTimeUtil.detectDateFormat(dateString);
		assertSame(detectedFormat, clientFormat);
	}

	/**
	 * Test method for
	 * {@link com.ihsinformatics.unittesting.DateTimeUtil#detectDateFormat(java.lang.String)}.
	 */
	@Test
	public void shouldNotDetectClientDateFormat() {
		String dateString = "1947-08-14 01:00:00";
		String detectedFormat = DateTimeUtil.detectDateFormat(dateString);
		assertNotSame(detectedFormat, clientFormat);
	}
}

File: unit-testing-workshop/src/python/datetime_util_test.py
import unittest
from datetime import datetime

from datetime_util import DateTimeUtil


class DateTimeUtilTest(unittest.TestCase):

    sql_format = "%Y-%m-%d %H:%M:%S"
    iso_format = "%Y-%m-%dT%H:%M:%S.Z"
    client_format = "%d/%m/%Y %H:%M"

    def test_should_convert_string_date_in_client_format(self):
        date = DateTimeUtil.from_string("01/01/2018 00:00", self.client_format)
        self.assertIsNotNone(date)

    def test_should_convert_string_datetime_in_client_format(self):
        date = DateTimeUtil.from_string("01/01/2018 16:55", self.client_format)
        self.assertIsNotNone(date)

    def test_should_convert_unusual_string_dates_in_client_format(self):
        unusual = [
            "-1/01/2018 00:00",
            "32/01/2018 00:00",
            "01/13/2018 00:00",
            "01/01/-800 00:00",
            "01/01/2018 24:00",
            "01/01/2018 00:60",
        ]
        for date_string in unusual:
            self.assertIsNotNone(
                DateTimeUtil.from_string(date_string, self.client_format),
                msg=f"Should not convert {date_string}",
            )

    def test_ambiguous_client_format_dates_from_string(self):
        date1 = DateTimeUtil.from_string("12/05/2018 00:00", self.client_format)
        date2 = DateTimeUtil.from_string("05/12/2018 00:00", self.client_format)
        self.assertTrue(date1 < date2)

    def test_should_convert_dates_to_client_format_string(self):
        dates = {
            datetime(2018, 5, 26, 0, 0): "26/05/2018 00:00",
            datetime(2018, 5, 26, 12, 30): "26/05/2018 12:30",
            datetime(2000, 1, 1, 0, 0): "01/01/2000 00:00",
        }
        for date, expected in dates.items():
            self.assertEqual(DateTimeUtil.to_string(date, self.client_format), expected)

    def test_from_sql_date_string(self):
        self.fail("Not yet implemented")

    def test_from_sql_datetime_string(self):
        self.fail("Not yet implemented")

    def test_to_sql_date_string(self):
        self.fail("Not yet implemented")

    def test_to_sql_datetime_string(self):
        self.fail("Not yet implemented")

    def test_should_throw_invalid_parameter_exception_on_unknown_format(self):
        date_string = "14081947 010000"
        with self.assertRaises(ValueError):
            DateTimeUtil.detect_date_format(date_string)

    def test_should_detect_sql_date_format(self):
        date_string = "1947-08-14 01:00:00"
        detected_format = DateTimeUtil.detect_date_format(date_string)
        self.assertEqual(detected_format, self.sql_format)

    def test_should_detect_client_date_format(self):
        date_string = "14/08/1947 01:00"
        detected_format = DateTimeUtil.detect_date_format(date_string)
        self.assertEqual(detected_format, self.client_format)

    def test_should_not_detect_client_date_format(self):
        date_string = "1947-08-14 01:00:00"
        detected_format = DateTimeUtil.detect_date_format(date_string)
        self.assertNotEqual(detected_format, self.client_format)


if __name__ == "__main__":
    unittest.main()
